### PR TITLE
QNN channel-last + semantic ArgMax performance improvements

### PR DESCRIPTION
**Pull request by glenn-jocher**
## Hexagon-friendly QNN graph I/O + in-graph semantic class maps

On-device profiling of the QNN integration (Snapdragon 8 Elite Gen 5 via the Flutter plugin, iPhone 17 Pro via the iOS SDK) identified I/O-level costs that the exported graph itself should solve. Both fixes are implemented as small `torch.nn.Module` export wrappers (the same pattern as `NMSModel`), replacing the earlier post-hoc ONNX graph surgery: a shared `ExportWrapper` base forwards attributes to the wrapped model, `QNNModel` handles input layout, and `ClassMapModel` handles semantic output.

### 1. Channel-last (NHWC) input — all QNN exports

The HTP's native layout is NHWC, so an NCHW-input graph pays a boundary Transpose on the NPU every inference — while apps simultaneously pay a CPU-side permute to *produce* NCHW from camera-native NHWC buffers. `QNNModel` prepends the permute in-graph; ORT's layout transformer folds it away during context generation, calibration is fed channel-last to match, and `QNNBackend` permutes its input when the session reports an NHWC graph. (Matches Qualcomm AI Hub's `--force_channel_last_input` recommendation.) Device result: detect inference 7.4 → 5.8 ms.

### 2. In-graph ArgMax class maps — QNN and CoreML semantic exports

Semantic exports previously emitted full float logits (~20M values at 1024px) that consumers dequantized and argmax-decoded on the CPU every frame. `ClassMapModel` appends `argmax(1)` + uint8 cast so the export emits a compact `[N, H, W]` class map (Qualcomm's published segmentation recipe). Semantic predict/val handle class-map outputs directly, so `yolo predict`/`yolo val` keep working with both forms.

Measured single-image burst latency (3 warmup + 15 timed runs, bus.jpg):

| Backend | Float logits | Class map |
| --- | --- | --- |
| QNN HTP, Snapdragon 8 Elite Gen 5 | 123–1065 ms (erratic) | **44 ms** (stable) |
| CoreML ANE, iPhone 17 Pro (A19) | 10.3 ms | **7.5 ms** (post 6.5 → 1.9 ms) |

### Tested and rejected (kept out by evidence)

- **Class maps for TFLite**: the GPU delegate cannot compile `ARG_MAX` with int64 indices (what onnx2tf emits; its argmax-replacement options no longer exist upstream), so LiteRT's whole-graph `CompiledModel` falls back to CPU — 137 ms vs 37.6 ms for GPU logits + consumer-side NHWC argmax. An unconditional wrapper also broke the NCNN semantic benchmark in CI. TFLite and the remaining formats keep logits; the wrapper is gated to `qnn`/`coreml`.
- **int32 class-map indices**: no benefit to offset its 4x larger NPU-to-CPU output transfer — every consumer reads uint8 (CoreML promotes it to int32 in-spec; ONNX Runtime reads it natively), so uint8 stays and int32 is reserved for >256-class models.

### Docs

`qnn.md` and `coreml.md` gain measured-performance tables (Snapdragon CPU/GPU/NPU and iPhone CPU/Neural Engine) with per-stage timing splits, burst-vs-sustained methodology disclosure, and tips linking the Android/iOS deployment repos; `coreml.md` modernized throughout.

### Version

Patch bump to 8.4.65 — deployment consumers pin model assets to the release that introduced this export format.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
This PR improves export and deployment support for mobile NPUs 📱⚡, especially Qualcomm QNN and Apple CoreML, while also tightening Docker build reliability, updating docs, and fixing several small edge-case bugs.

### 📊 Key Changes
- Added new export-time wrappers in `exporter.py` for smarter deployment:
  - `QNNModel` enables channel-last input for Qualcomm QNN exports 🧠
  - `ClassMapModel` makes semantic segmentation exports output compact class maps instead of large raw logits
  - `ExportWrapper` provides shared wrapper behavior for export logic
- Improved **QNN export pipeline**:
  - QNN exports now use NHWC/channel-last layout, which better matches Snapdragon hardware and camera pipelines
  - QNN backend now detects and handles channel-last inputs automatically
  - QNN quantization preprocessing was updated to support this new layout
- Improved **semantic segmentation export** for **QNN** and **CoreML**:
  - Semantic models can now export class maps directly, reducing CPU-side postprocessing
  - Prediction and validation code was updated to correctly handle both logits and precomputed class maps
  - `scale_masks()` now supports nearest-neighbor interpolation to avoid blending segmentation labels 🎯
- Updated **export docs** to reflect modern YOLO26 behavior:
  - Clarified end-to-end detection outputs for YOLO26
  - Updated export argument docs, including `data=None` default behavior and RKNN `half` support
  - Added new reference docs for export helper classes
- Expanded **CoreML** and **QNN** integration docs significantly 📘:
  - Added clearer deployment guidance
  - Added measured mobile performance tables
  - Added recommendations for iOS, Flutter, Neural Engine, and Snapdragon NPU workflows
- Improved **Docker build workflow and images** 🐳:
  - Added `--no-cache` package installs in Dockerfiles
  - Added repeated `docker builder prune -af` cleanup to reduce builder-state issues
  - Removed an extra workflow trigger step that launched downstream actions
  - Improved shell quoting in GitHub Actions for safer workflow execution
- Added/updated **tests**:
  - RKNN export is now tested for both `int8=True` and `int8=False`
- Fixed several smaller issues and doc inconsistencies:
  - Safer empty-result handling in the TFLite example
  - Action recognition example now avoids invalid `topk(2)` calls when fewer labels exist
  - Tracking docs now mention OBB support
  - Solutions docs now reflect that some workout fields may return per-person lists
  - Multiple README, examples, comments, and docstring corrections ✨

### 🎯 Purpose & Impact
- **Faster mobile deployment**: QNN and CoreML exports are now better optimized for real mobile hardware, especially Snapdragon NPUs and Apple Neural Engine 🚀
- **Lower postprocessing overhead**: Exporting semantic class maps directly can reduce latency and improve consistency for on-device segmentation
- **Better developer experience**: Clearer docs, updated references, and corrected examples make export/deployment workflows easier to understand and use
- **More reliable CI and Docker builds**: Cleanup steps and safer install behavior should reduce flaky image builds and workflow issues
- **Improved robustness**: Small bug fixes help avoid crashes or confusing behavior in edge cases, especially in example apps and export/testing flows
- **Broader user benefit**: App developers, edge deployers, and teams using YOLO26 for mobile inference should see smoother setup, clearer guidance, and better performance 📱✅

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @glenn-jocher, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help with review readiness; an engineer will assist shortly. Please review the checklist below for a smooth integration. 😊

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/utils/export/qnn.py [hunk 1]
```diff
@@ -34,6 +34,61 @@ def qnn_library_paths() -> tuple[str | None, str]:
...
+    """
+    import onnx
+    from onnx import TensorProto, helper
+
+    model = onnx.load(str(onnx_file))
+    graph = model.graph
+
+    nhwc = False
+    inp = graph.input[0]
+    dims = [d.dim_value for d in inp.type.tensor_type.shape.dim]
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/export/qnn.py]: ⚠️ **HIGH**: This rewrite only reads and writes `dim_value`, so any symbolic dimensions from the exported ONNX (for example dynamic batch) are lost. A dynamic input like `[batch, 3, H, W]` gets rewritten to `[0, H, W, 3]`, and the semantic output shape below has the same problem. Please preserve each dimension's `dim_param` when present instead of rebuilding the shape from raw `dim_value`s.

File: ultralytics/utils/export/qnn.py [hunk 2]
```diff
@@ -34,6 +34,61 @@ def qnn_library_paths() -> tuple[str | None, str]:
...
+        for d in (n, h, w, c):
+            inp.type.tensor_type.shape.dim.add().dim_value = d
+        nhwc = True
+
+    if task == "semantic":
+        out = graph.output[0]
+        odims = [d.dim_value for d in out.type.tensor_type.shape.dim]
+        graph.node.extend(
+            [
+                helper.make_node("ArgMax", [out.name], [f"{out.name}_argmax"], axis=1, keepdims=0, name="qnn_argmax"),
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/export/qnn.py]: ⚠️ **HIGH**: `ArgMax(axis=1)` breaks binary semantic models. When the logits have shape `[N, 1, H, W]`, argmax over the single channel is always 0, so every exported prediction becomes background. This needs the same special-case behavior the PyTorch path uses (`> 0`) or the in-graph rewrite should be skipped for single-channel outputs.

File: ultralytics/models/yolo/semantic/predict.py [hunk 3]
```diff
@@ -58,13 +58,17 @@ def postprocess(self, preds, img, orig_imgs):
         results = []
         for i, (pred, orig_img) in enumerate(zip(preds, orig_imgs)):
             img_path = self.batch[0][i] if isinstance(self.batch[0], list) else self.batch[0]
-            pred = pred.unsqueeze(0).float()
+            class_map_input = pred.ndim == 2  # exports with in-graph ArgMax emit a [H, W] class map directly
```
Comment (UltralyticsAssistant) [inline ultralytics/models/yolo/semantic/predict.py:61]: 💡 **MEDIUM**: The new direct-class-map branch still passes through `ops.scale_masks()` later in the function, which resizes with bilinear interpolation. That is fine for logits, but it blends discrete class IDs and produces incorrect labels before the final cast. Exported `[H, W]` class maps need a nearest-neighbor resize/crop path all the way through postprocessing.

File: ultralytics/utils/export/qnn.py [hunk 4]
```diff
@@ -34,6 +34,61 @@ def qnn_library_paths() -> tuple[str | None, str]:
...
+                helper.make_node("Cast", [f"{out.name}_argmax"], ["class_map"], to=TensorProto.UINT8, name="qnn_classmap"),
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/export/qnn.py]: 💡 **MEDIUM**: Hard-casting every semantic class map to `UINT8` will silently corrupt labels for models with more than 255 classes. Since the predictor/validator already support wider integer dtypes, the export path should choose the cast type from the output channel count or avoid narrowing when `nc > 256`.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Nice feature addition overall, and the export/backend plumbing is coherent. I found a few correctness issues to address before merging: dynamic ONNX dimensions are not preserved during the graph rewrite, binary semantic models will export incorrect all-zero masks, `uint8` casts can overflow for larger class counts, and direct class-map outputs are still being bilinearly scaled during prediction.

💬 Posted 4 inline comments

**Comment by glenn-jocher** [comment]
**Device + Python validation complete**:

- **NHWC detect** (Snapdragon 8 Elite Gen 5): inference 7.4→5.8 ms (boundary transpose eliminated), end-to-end 11.5 ms with the Flutter plugin's flat decode — vs 23.3 ms at yesterday's baseline. Detections correct (bus + person).
- **In-graph-ArgMax semantic**: **49.7 ms stable** on the NPU vs 123-1065 ms (erratic) for the logits export — the variance disappears with the 80 MB float output. Valid class map confirmed on-device.
- **Python environment**: both new-format models run correctly through `yolo predict` (QNNBackend NHWC permute + semantic class-map handling) on the x86-64 HTP simulator: detect reports '4 persons, 1 bus'.

**Comment by glenn-jocher** [comment]
**Final device benchmarks with this PR's exports** (Snapdragon 8 Elite Gen 5, all six tasks, single-image; full table updated in qnn.md):

| Task | QNN before (NCHW logits) | QNN with this PR | 
|---|---|---|
| Detect | 23.3 ms | **11.3 ms** |
| Segment | 23.5 ms | **21.3 ms** |
| Semantic | 123-1065 ms (erratic) | **49.1 ms (stable)** |
| Classify | 2.3 ms | 2.4 ms |
| Pose | 12.3 ms | **10.8 ms** |
| OBB | 26.2 ms | **21.0 ms** |

Correctness on-device: detect/segment find bus+person, classify top-1 `police_van`, pose finds persons, semantic produces a valid class map. Python environment verified via the x86-64 HTP simulator. Version bumped to 8.4.65. After merge, the mobile release assets get re-exported in this format.

**Comment by glenn-jocher** [comment]
Refactored per review discussion: the I/O adaptation is now a **torch `QNNModel` wrapper** (transpose at the input + `argmax(1).byte()` for semantic in `forward`), traced by the standard ONNX export — the same pattern as `NMSModel` and the CoreML export wrapper — replacing the `_prepare_qnn_graph` ONNX protobuf surgery (net −40 lines). `onnx2qnn` detects channel-last graphs from the input shape to feed calibration correctly.

Re-verified after the refactor: exported context binaries have identical I/O (detect `[1,640,640,3]` input; semantic uint8 `[1,1024,1024]` class map) and the Python pipeline produces identical results ('4 persons, 1 bus') — so the device benchmarks in the table stand unchanged.

**Comment by glenn-jocher** [comment]
Confirmation benchmark with all six models exported from the final `QNNModel`-wrapper code (Snapdragon 8 Elite Gen 5): detect 11.6 / segment 21.9 / semantic 49.4 / classify 2.6 / pose 11.0 / OBB 21.1 ms — within run-to-run noise (≤0.6 ms) of the published table, all correctness checks passing. The refactor is performance-neutral as expected; the qnn.md numbers stand.

File: ultralytics/utils/export/qnn.py (hunk 1 above)
Comment (glenn-jocher) [reply ultralytics/utils/export/qnn.py]: The ONNX graph rewrite was replaced by torch-level export wrappers (`QNNModel` permutes input in `forward`), so there is no dims surgery left — symbolic dims never arise (QNN context binaries are fixed-shape by design and `batch` is pinned at export).

File: ultralytics/utils/export/qnn.py (hunk 2 above)
Comment (glenn-jocher) [reply ultralytics/utils/export/qnn.py]: Fixed in 26d28b674: `ClassMapModel` now thresholds single-channel logits (`y[:, 0].gt(0)`) instead of argmax, matching predict/val `nc == 1` semantics.

File: ultralytics/models/yolo/semantic/predict.py (hunk 3 above)
Comment (glenn-jocher) [reply ultralytics/models/yolo/semantic/predict.py:61]: Fixed in 47a54c22b: `scale_masks` gained a `mode` parameter and the class-map branch passes `nearest`, so label IDs are never blended (verified: upscaled maps contain only original IDs).

File: ultralytics/utils/export/qnn.py (hunk 4 above)
Comment (glenn-jocher) [reply ultralytics/utils/export/qnn.py]: Fixed in 47a54c22b: the cast dtype is now selected per model — uint8 up to 256 classes (keeps the whole graph on the HTP; int32 measured 1054 ms vs 44 ms there), int32 beyond that.

**Comment by glenn-jocher** [comment]
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

GitHub Actions below triggered via workflow dispatch for this PR at 2026-06-11 00:30:50 UTC with `@ultralytics/run-all` command
(available commands are `@ultralytics/run-all`, `@ultralytics/run-ci`, and `@ultralytics/run-docker`):

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/27315481534): `ci.yml` run 23754
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/27315482384): `docker.yml` run 2350

File: ultralytics/utils/export/qnn.py (hunk 4 above)
Comment (glenn-jocher) [reply ultralytics/utils/export/qnn.py]: Correction to the timing cited above: the 1054 ms artifact turned out to be an old-format logits export, not an int32 class map. The dtype choice stands on output size — uint8 quarters the NPU-to-CPU transfer and every consumer reads it natively — with int32 reserved for >256-class models.

**Comment by UltralyticsAssistant** [comment]
🎉 Excited to see this merged — fantastic work by @glenn-jocher with valuable contributions from @Laughing-q!

> “Great things are done by a series of small things brought together.” — Vincent van Gogh

This PR is a great example of that idea in action: smarter QNN and CoreML export behavior, more efficient segmentation outputs, stronger Docker and CI reliability, clearer YOLO26 docs, and a collection of thoughtful edge-case fixes that together make mobile deployment smoother and more robust for everyone.

Thanks for pushing forward the developer experience and real-world deployment story for YOLO26 across Apple and Snapdragon devices 📱⚡
